### PR TITLE
Scaffolding for symbolic AD in the JIT

### PR DESCRIPTION
**Pull request by apaszke**

### CODE REVIEW

File: torch/csrc/jit/test_jit.cpp [hunk 1]
```diff
@@ -496,7 +435,122 @@ void interpStageTest() {
...
+  static const std::vector<ADTestSpec> ad_tests = {
+    {"add", binary_pointwise, [](const VL& v) -> VL { return {v[0] + v[1]}; }},
+    {"sub", binary_pointwise, [](const VL& v) -> VL { return {v[0] - v[1]}; }},
+    {"mul", binary_pointwise, [](const VL& v) -> VL { return {v[0] * v[1]}; }},
+  };
+
+  // We have to release the GIL inside this method, because if we happen to
+  // initialize the autograd engine here, the newly spawned worker threads will
+  // try to initialize their PyThreadState*, and they need the GIL for this.
+  AutoNoGIL _no_gil;
```
Comment (ezyang) [inline torch/csrc/jit/test_jit.cpp]: Why aren't we releasing the GIL at the top level of the tests?

Comment (apaszke) [reply torch/csrc/jit/test_jit.cpp]: ¯\_(ツ)_/¯ I didn't make it test-global just in case any other tests need it

Comment (apaszke) [reply torch/csrc/jit/test_jit.cpp]: But in general it doesn't really matter

File: torch/csrc/jit/autodiff.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#include "torch/csrc/jit/autodiff.h"
+
+#include "torch/csrc/jit/symbolic_variable.h"
+#include "torch/csrc/utils/functional.h"
+
+namespace torch { namespace jit {
+
+std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_values) {
+  const auto build_sym_grad = [node](const std::vector<SymbolicVariable>& grads) -> std::vector<SymbolicVariable> {
```
Comment (zdevito) [inline torch/csrc/jit/autodiff.cpp]: I think inlining this lambda would be easier to understand,  since it is only used once.

File: torch/csrc/jit/autodiff.h [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#pragma once
```
Comment (zdevito) [inline torch/csrc/jit/autodiff.h]: nit: autodiff traditionally means tape-and-recording stuff. So we should probably call this `symbolic_diff.h`

File: torch/csrc/jit/autodiff.cpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    switch(node->kind()) {
+      case kadd:
+        return {grads[0], grads[0]};
+      case ksub:
+        return {grads[0], -grads[0]};
+      case kmul:
+        auto inputs = node->inputs();
```
Comment (zdevito) [inline torch/csrc/jit/autodiff.cpp]: nit: if we add this line before the switch, then any op that needs the inputs won't have to repeat this line, we could do the same for outputs as well. That way, we can keep many ops 1 liners.

File: torch/csrc/jit/test_jit.cpp [hunk 5]
```diff
@@ -496,7 +435,122 @@ void interpStageTest() {
     JIT_ASSERT(exactlyEqual(outputs[1],cx));
 }
 
+using var_meta_type = std::vector<int64_t>;
```
Comment (zdevito) [inline torch/csrc/jit/test_jit.cpp]: nit: var_size_spec or something like that "meta_type" suggests some something more magical than this is :)

File: torch/csrc/jit/test_jit.cpp [hunk 6]
```diff
@@ -496,7 +435,122 @@ void interpStageTest() {
...
+
+void assertAllClose(const tensor_list& a, const tensor_list& b) {
+  JIT_ASSERT(a.size() == b.size());
+  for (std::size_t i = 0; i < a.size(); ++i) {
+    JIT_ASSERT(a[i].is_same_size(b[i]));
+    JIT_ASSERT(a[i].allclose(b[i]));
+  }
+}
+
+void testAD() {
```
Comment (zdevito) [inline torch/csrc/jit/test_jit.cpp]: This seems good! Simple but will keep this synced with autograd.

**Review by zdevito — APPROVED** [review]
Looks good. I just have some nits for clarity.

File: torch/csrc/jit/autodiff.cpp (hunk 2 above)
Comment (apaszke) [reply torch/csrc/jit/autodiff.cpp]: I agree, but it would make the code 2x longer, because I couldn't use `return` 😉 I like it more this way, but don't mind inlining it if you'd prefer that way

File: torch/csrc/jit/autodiff.cpp [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    grad_map[output] = graph->addInput()->setType(output->typeOption());
+
+  for (auto it = graph->rbegin(), end = graph->rend(); it != end; ++it) {
+    Node *node = *it;
+    auto inputs = node->inputs();
+    value_list grad_inputs = gradientForNode(node, fmap(node->outputs(), get_grad));
+    JIT_ASSERT(grad_inputs.size() == node->inputs().size());
+    for (std::size_t i = 0, num_inputs = grad_inputs.size(); i < num_inputs; ++i) {
+      if (Value * prev_grad = grad_map[inputs[i]]) {
+        Node *new_grad_node = graph->create(kadd, {prev_grad, grad_inputs[i]})
```
Comment (zdevito) [inline torch/csrc/jit/autodiff.cpp]: Maybe use the SymbolicVariable to keep this short: SymbolicVariable(prev_grad) + SymbolicVariable(grad_inputs[i])

File: torch/csrc/jit/autodiff.cpp [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        return {grads[0] * inputs[1], grads[0] * inputs[0]};
+    }
+    throw std::runtime_error(std::string("don't support differentiation of `") +
+                            node->kind().toString() + "`");
+  };
+  auto sym_grads = build_sym_grad(fmap<SymbolicVariable>(grad_values));
+  return fmap(sym_grads, [](const SymbolicVariable &v) { return v.value(); });
+}
+
+void differentiate(std::shared_ptr<Graph>& graph) {
```
Comment (zdevito) [inline torch/csrc/jit/autodiff.cpp]: We will need a version that takes  `, std::vector<bool> requires_grad)`, feel free to do that in a later PR, or we can add it when we run into it.

File: torch/csrc/jit/autodiff.cpp [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    for (Value *input : node->inputs()) {
+      // XXX: this assumes that the single Param node in the graph (that holds inputs) has stage 0
+      if (input->node()->stage() == 0)
+        backward_inputs_set.insert(input);
+    }
+  }
+  value_list backward_inputs(backward_inputs_set.begin(), backward_inputs_set.end());
+  // We want them nicely sorted with inputs being first, and then following
+  // the topological ordering (at least approximately, we could do a better job here).
+  std::sort(backward_inputs.begin(), backward_inputs.end(),
```
Comment (zdevito) [inline torch/csrc/jit/autodiff.cpp]: Let's do this in true topological order.  Having a confusing order for the captures will make it hard to debug.

File: torch/csrc/jit/autodiff.cpp [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    }
+  }
+
+  for (auto input : graph->inputs()) {
+    if (input->stage() > 0) break;
+    graph->registerOutput(grad_map.at(input));
+  }
+}
+
+// TODO: return metadata that allows to associate inputs with values form graph
```
Comment (zdevito) [inline torch/csrc/jit/autodiff.cpp]: + comment about what this does, what the inputs to the graph will be.

File: torch/csrc/jit/autodiff.cpp [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    Node *clone = dgraph->createClone(node, lookup_val);
+    for (std::size_t i = 0, num_outputs = clone->outputs().size(); i < num_outputs; ++i)
+      val_map[node->outputs()[i]] = clone->outputs()[i];
+    dgraph->appendNode(clone);
+  }
+  for (auto output : graph->outputs()) {
+    if (output->stage() == 0) continue;
+    dgraph->registerOutput(val_map.at(output));
+  }
+  return std::make_pair(dgraph, backward_inputs);
```
Comment (zdevito) [inline torch/csrc/jit/autodiff.cpp]: For each Value in `backward_inputs` that is not already in the input or output list of `graph`, we need to add an additional output to `graph`. I think it is appropriate for this function to do this as well.

File: torch/csrc/jit/autodiff.cpp (hunk 8 above)
Comment (zdevito) [inline torch/csrc/jit/autodiff.cpp]: comment about what this code does (e.g. what inputs and outputs will it add?)

File: torch/csrc/jit/test_jit.cpp [hunk 12]
```diff
@@ -496,7 +435,168 @@ void interpStageTest() {
...
+  }
+}
+
+std::string toString(std::shared_ptr<Graph>& graph) {
+  std::ostringstream s;
+  s << *graph;
+  return s.str();
+}
+
+void testGraphDiff() {
```
Comment (zdevito) [inline torch/csrc/jit/test_jit.cpp]: We will need more rigorous testing for boundary cases, but that can wait until we have the meta-data for gradient, so that we can actually test by running the code.

**Review by zdevito — COMMENTED** [review]
Looked at the differentiate and lambdaLift functions as well. Nice and simple. I had a few comments, but no major changes.

File: torch/csrc/jit/autodiff.cpp (hunk 11 above)
Comment (zdevito) [reply torch/csrc/jit/autodiff.cpp]: Also, for each output `t` added to `graph` as described above, we need to add a `dt` input to `dgraph`, and  patch the gradient computation for `t` so that it starts set to the `dt`. This temporary only ever gets  used if a gradient of this gradient exists, but we always have to create it here for simplicity. I tried other designs that did attempted to defer putting these in, but it is easiest to do when originally calculating the gradient, because you already have the meta-data in the grad_map. This won't change the perf. because the optimizer will recognize that `dt` must be 0 and remove it during specialization of the gradient.

File: torch/csrc/jit/autodiff.cpp (hunk 8 above)
Comment (apaszke) [reply torch/csrc/jit/autodiff.cpp]: I wasn't sure how exactly do we want to represent that information, so I omitted that. I'll add it in a later PR so we don't have to go over all this code again

File: torch/csrc/jit/autodiff.cpp (hunk 9 above)
Comment (apaszke) [reply torch/csrc/jit/autodiff.cpp]: I'm a bit worried that we'll keep reconstructing these topo_idx maps over and over (they appear in 3 places now I think). On the other hand maintaining an ordering on a changing graph is annoying, but possible if only we had a simple balanced tree implementation. I don't think it can be implemented using `std`, so we'd need an extra dependency or to roll our own

File: torch/csrc/jit/autodiff.cpp (hunk 11 above)
Comment (apaszke) [reply torch/csrc/jit/autodiff.cpp]: Hmm good point. This lifting has to give us two graphs, not one.

File: torch/csrc/jit/test_jit.cpp (hunk 12 above)
Comment (apaszke) [reply torch/csrc/jit/test_jit.cpp]: Yeah that was mostly meant as a smoke test so you can see what the output is

**Comment by apaszke** [comment]
I've decided to simplify the patch by only including the scaffolding that you approved. I'll improve the lambda lifting and send another PR with that.
